src/mon/OSDMonitor.cc

```diff
@@ -1127,6 +1127,48 @@ void OSDMonitor::encode_pending(MonitorDBStore::TransactionRef t)
 			<< "required " << ceph_release_name(mv);
       pending_inc.new_require_min_compat_client = mv;
     }
+
+    // upgrade to mimic?
+    if (osdmap.require_osd_release < CEPH_RELEASE_MIMIC &&
+	tmp.require_osd_release >= CEPH_RELEASE_MIMIC) {
+      dout(10) << __func__ << " first mimic epoch" << dendl;
+      // record this epoch as the deletion for all legacy removed_snaps
+      for (auto& p : tmp.get_pools()) {
+	// update every pool
+	if (pending_inc.new_pools.count(p.first) == 0) {
+	  pending_inc.new_pools[p.first] = p.second;
+	}
+	auto& pi = pending_inc.new_pools[p.first];
+	if (pi.snap_seq > 0 &&
+	    (pi.flags & (pg_pool_t::FLAG_SELFMANAGED_SNAPS |
+			 pg_pool_t::FLAG_POOL_SNAPS)) == 0) {
+	  if (!pi.removed_snaps.empty()) {
+	    pi.flags |= pg_pool_t::FLAG_SELFMANAGED_SNAPS;
+	  } else {
+	    pi.flags |= pg_pool_t::FLAG_POOL_SNAPS;
+	  }
+	}
+
+	if (!p.second.removed_snaps.empty()) {
+	  // all snaps removed this epoch
+	  dout(10) << __func__ << " converting pool " << p.first
+		   << " with " << p.second.removed_snaps.size()
+		   << " legacy removed_snaps" << dendl;
+	  string k = make_snap_epoch_key(p.first, pending_inc.epoch);
+	  bufferlist v;
+	  ::encode(p.second.removed_snaps, v);
+	  t->put(OSD_SNAP_PREFIX, k, v);
+	  for (auto q = p.second.removed_snaps.begin();
+	       q != p.second.removed_snaps.end();
+	       ++q) {
+	    bufferlist v;
+	    string k = make_snap_key_value(p.first, q.get_start(),
+					   q.get_len(), pending_inc.epoch, &v);
+	    t->put(OSD_SNAP_PREFIX, k, v);
```

Comment:
only snaps removed >= mimic are added to new_removed_snaps.  prior to that the OSDs are calculating that themselves by comparing the pg_pool_t removed_snaps interval_set like before.

Reply:
What about if we removed snaps in this map epoch prior to setting the mimc flag?

And I must be dense, but I'm just not seeing how the OSD or OSDMap put stuff into the new places automatically. It looks like we clear the snap_trimq and start setting it equal to the removed_snaps_queue right away in the PG code.

Also, for purged snaps, we'll start with the beginning and we're only doing 100 intervals at a time. It may take quite a lot of maps before we catch up to the cutover point, and until we do the removed_snaps_queue member will be growing linearly. But I guess that's okay since it's just another copy of data we would have had pre-mimic anyway.

Reply:
Hmm you're right, this isn't quite right.  Fixing it up!